### PR TITLE
Adding a Neo4j GIE

### PR DESCRIPTION
**Pull request by thobalose**
The Neo4j IE described in this [issue](https://github.com/bxlab/galaxy_hackathon/issues/26).

Thanks to @bgruening and @erasche.

### CODE REVIEW

File: lib/galaxy/web/proxy/__init__.py [hunk 1]
```diff
@@ -106,6 +106,7 @@ def launch_proxy_command(self, config):
             "--sessions", config.proxy_session_map,
             "--ip", config.dynamic_proxy_bind_ip,
             "--port", str(config.dynamic_proxy_bind_port),
+            "--cookie", "galaxysession",
```
Comment (hexylena) [inline lib/galaxy/web/proxy/__init__.py]: is this needed? That should be default.

File: config/plugins/interactive_environments/neo/config/neo.ini.sample [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+#
+# It is recommended that you use command_inject if you need to inject
+# additional parameters. This command string is re-used for a `docker inspect`
+# command and will likely cause errors if it is extensively modified, past the
+# usual group/sudo changes.
+#command = docker {docker_args}
+
+# The image argument was moved to "allowed_images.yml.sample"
+
+# Additional arguments that are passed to the `docker run` command.
```
Comment (pvanheus) [inline config/plugins/interactive_environments/neo/config/neo.ini.sample:20]: Add examples of what NEO4J_UID and NEO4J_GID variables should be set to.

File: config/plugins/interactive_environments/neo/templates/neo.mako [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    requirejs(['interactive_environments', 'plugin/neo'], function () {
+        load_notebook(url);
+    });
+
+</script>
+<div id="main" width="100%" height="100%">
+</div>
+</body>
+</html>
```
Comment (pvanheus) [inline config/plugins/interactive_environments/neo/templates/neo.mako:43]: Add end of line

File: lib/galaxy/web/proxy/__init__.py (hunk 1 above)
Comment (thobalose) [inline lib/galaxy/web/proxy/__init__.py]: This must have been whilst debugging the issues with proxy_ipc. Having [seen](https://docs.galaxyproject.org/en/master/admin/interactive_environments.html#setting-up-the-proxy) and noting that it's not present when galaxy launched the proxy.

File: .gitignore [hunk 4]
```diff
@@ -129,3 +129,4 @@ doc/build
 .DS_Store
 *.rej
 *~
+.idea/
```
Comment (pvanheus) [inline .gitignore:132]: Do we need this change? If we do, add end of line.

**Comment by pvanheus** [comment]
For this code to work with the nginx proxy it needs this in the nginx.conf:

```
        #Neo4J
        location /gie_proxy/neo {
            proxy_pass http://localhost:8800/;
            proxy_redirect default;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
        }

        location /db/data {
            proxy_pass http://localhost:8800/db/data;
            proxy_redirect default;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
        }
```
1. Does the first block clash with other IEs? I.e. will the proxy know to direct that message to the correct IE?
2. The second block does not refer to /gie_proxy. This is because the AngularJS app that implements the Neo4j shell does calls to that URL - this is not configurable at present, but a PR should be made against the Neo4j code to make it configurable.

If this PR is being merged, we need to add this info the GIE docs.

**Comment by pvanheus** [comment]
A bit of testing suggests that (1) is causing an issue. To reproduce the problem:
1. Start a Jupyter IE
2. Start a Neo4j IE
3. The Jupyter IE goes to "Not found" and control of the IE is lost from the side of Galaxy.

In other words, this IE doesn't play nicely with other IEs on the same server.

**Comment by bgruening** [comment]
As discussed this is a general IE issue at the moment, at least with the IE proxy, afaik.

**Pull request closed without merging**